### DISCUSSION TITLE
How to disable check for renames?

### DISCUSSION THREAD

**Discussion by dudicoco** [gitpython-developers/GitPython#1195] [Q&A] [answered]
Hi,

I've tried to use the following to disable git diff renamed files search:
```
repo.index.diff('origin/HEAD', no_renames=True)
```

This did not work, and after some digging, i've found that GitPython hardcodes the `-M` flag:
https://github.com/gitpython-developers/GitPython/blob/master/git/diff.py#L106

Why is this flag hardcoded and is there a way around it?

Thanks

**Comment by Byron** [comment, marked as answer]
With this API there is no way around it unless the git command is used directly such as in `repo.git.diff(…)`.
If there is motivation, a PR could be added to disable rename tracking which I presume is in the interest of increased performance.

**Reply by dudicoco** [reply]
@Byron there are two issues here:
1. Sometimes it is necessary to to add `no-renames` to prevent false positives in git's rename detection - for example, deleting a file and creating another file with similar content (this will be detected as rename, rather than delete + add).
2. It seems that there is a bug GitPython's diff filter when used with renames. To recreate, delete a file and create another file with similar contents, compare the result of `git diff --name-only --diff-filter=M -M` with GitPython's `iter_change_type('M')`. GitPython will show the deleted/renamed file as modified.

**Comment by mellowed100** [comment]
Hello! I've hit this issue as well. I'd like to try the repo.git.diff() approach as a workaround. Using the results from repo.git.diff(), is there an easy way to build a DiffIndex()?
Thanks!

**Reply by Byron** [reply]
I don't think there is an 'easy' way, but I am sure one can do what it takes to make `-M` configurable. A contribution for this would certainly be appreciated as it would solve a real problem.

**Reply by mellowed100** [reply]
I can give it a try. I know adding kwargs like **no_renames=True** will produce **--no-renames**. Is it possible to add a kwarg to produce something like **-M75%**?
thanks!

**Reply by Byron** [reply]
For this to work, one would have to teach the diff-parser to handle renames, so that `--no-renames` doesn't need to be forcefully passed anymore. Then it's possible to add any *kwarg* one desires, like `M='75%'`.

**Reply by mellowed100** [reply]
Agree. I was trying to do something simple like replace:

`args.append("-M")  # check for renames, in both formats
`

with:

```
if not any(x in kwargs for x in ("no_renames","M")):
   args.append("-M")  # check for renames, in both formats

```

This would leave the default behavior the same. 

Adding **no_renames** would work too, by swapping out **-M** with **--no-renames**. 
`commitDiff = commitA.diff(commitB, no_renames=True)`

Where I get hung up is on creating the **-M75%** arg. Having a call like this:

`commitDiff = commitA.diff(commitB, M='75%)`

This will generate a call like this:
`git diff-tree -M 75% 69a8e0b10c8695d7f073c3b32c647aa8e892b1e2 140ef4c33fa5a899cd8d1dfb15a03c342a49a95c -r --abbrev=40 --full-index --raw -z --no-color`

Where git is looking for something like this:
`git diff-tree -M75% 69a8e0b10c8695d7f073c3b32c647aa8e892b1e2 140ef4c33fa5a899cd8d1dfb15a03c342a49a95c -r --abbrev=40 --full-index --raw -z --no-color`

I'm not clear on how to setup the **M="75%"** arg to produce:
**-M75%**
instead of
**-M 75%**
which it is doing now. Git doesn't seem to like that extra space in there. I was hoping there was some way to indicate to the kwargs converter that it shouldn't insert that space in there.

**Reply by Byron** [reply]
> I'm not clear on how to setup the M="75%" arg to produce:

This should work like this: `commitA.diff(commitB, '-M75%')`

**Reply by mellowed100** [reply]
Hey Bryon,
Thank you for the suggestion. Unfortunately, `Diffable.diff()` treats '-M75%' as a path, and not as an argument. Here is what get produced:

`git diff-tree 58259d006066372eb337cce81caf97365eb30f1e 22315b3c433865970e0a2fcd1b599786c2e01703 -r --abbrev=40 --full-index -M --raw -z --no-color -- -M75%`

It ends up putting `'-M75%'` after `'--'`. 
I could probably extract a kwarg like `M='75%'` and manually convert that to `'-M75%'`, but I was hoping there was an existing mechanism for that.  

thanks!

**Reply by mellowed100** [reply]
```
        # remove -M arg if user supplies 'no_renames' or 'M' kwargs 
        if not any(x in kwargs for x in ('no_renames','M')):
            args.append("-M")  # check for renames, in both formats
        
        # manually convert '-M'
        if 'M' in kwargs:
            args.append(f"-M{str(kwargs['M'])}")
            del(kwargs['M'])
            
```



`commitA.diff(commitB, M='75%')`
produces:
`git diff-tree 58259d006066372eb337cce81caf97365eb30f1e 22315b3c433865970e0a2fcd1b599786c2e01703 -r --abbrev=40 --full-index -M75% --raw -z --no-color`




`commitA.diff(commitB, no_renames=True)`
produces:
`git diff-tree --no-renames 58259d006066372eb337cce81caf97365eb30f1e 22315b3c433865970e0a2fcd1b599786c2e01703 -r --abbrev=40 --full-index --raw -z --no-color`




default behavior stays the same
`commitA.diff(commitB)`
produces:
git diff-tree 58259d006066372eb337cce81caf97365eb30f1e 22315b3c433865970e0a2fcd1b599786c2e01703 -r --abbrev=40 --full-index -M --raw -z --no-color

**Reply by Byron** [reply]
Maybe that could be generalized (and tested) to become a contribution - it looks like a genuine improvement without adding too much complexity.

**Reply by mellowed100** [reply]
Hey Bryon,

So I took another look into the command line formatting issue. It turns out there is an option to format those single char kawrgs.
**split_single_char_options** controls the inclusion of a space character when constructing the git command line.

original:
commitA.diff(commitB, M='75%')

produces the **-M 75%:**  (space included)
git diff-tree -M 75% 69a8e0b10c8695d7f073c3b32c647aa8e892b1e2 140ef4c33fa5a899cd8d1dfb15a03c342a49a95c -r --abbrev=40 --full-index --raw -z --no-color

Including option to format single char kwargs:
commitA.diff(commitB, **split_single_char_options=False** ,M='75%')
produces the **-M75%** (space removed)
git diff-tree -M75% 69a8e0b10c8695d7f073c3b32c647aa8e892b1e2 140ef4c33fa5a899cd8d1dfb15a03c342a49a95c -r --abbrev=40 --full-index --raw -z --no-color

So it looks like all we need is to remove the default **-M** when the user wants to override it with either **--find-renames=<n>** or **-M<n>** or **--no-renames**

        # remove default '-M' arg (check for renames) if user is overriding it
        if not any(x in kwargs for x in ('find_renames', 'no_renames','M')):
            args.append("-M")

The remaining kwargs can be processed through the normal routines.

**Reply by mellowed100** [reply]
It may also be worth adding the **split_single_char_options** option in the docs.

**Reply by Byron** [reply]
I see, thanks so much for investigating this. A PR to improve the situation would definitely be welcome. Thank you.

**Comment by nwcm** [comment]
would also like for this to be added!

**Reply by mellowed100** [reply]
Hey nwcm! The PR was finished last week. You can probably download the source now, or wait for the official 3.1.31 release.
https://github.com/gitpython-developers/GitPython/pull/1551
Usage examples can be found in test/test_diff.py test_rename_override().
